modules/system/defaults/dock.nix

```diff
@@ -131,9 +131,16 @@ in {
         Persistent applications in the dock.
       '';
       apply = value:
+        let
+          tileTypes = ["spacer-tile" "small-spacer-tile"];
+        in
```

Comment:
Maybe this could be made even more readable with some named functions?

```nix
apply =
let
  tileTypes = ["spacer-tile" "small-spacer-tile"];
  toSpecialTile = type: { tile-data = {}; tile-type = type; };
  toAppTile = cfurl: { tile-data = { file-data = { _CFURLString = cfurl; _CFURLStringType = 0; }; }; };
  toTile = s: if elem s tileTypes then toSpecialTile s else toAppTile s;
in
value: if isList value then map toTile value else value;
```

However, seeing as this is being done to the option value, why not take advantage of `types.coercedTo` to a) allow users to define custom tile attrsets and b) simplify the coercion?

`coercedTo` also improves documentation by being explicit about the fact that coercion is done. Coercing manually in `apply` can surprise users who expect to be able to _read_ the same value types that they _define_.

```nix
type =
let
  # The same helpers as above
  specialTileTypes = ["spacer-tile" "small-spacer-tile"];
  toSpecialTile = type: { tile-data = {}; tile-type = type; };
  toAppTile = cfurl: { tile-data = { file-data = { _CFURLString = cfurl; _CFURLStringType = 0; }; }; };
  
  # The coercion function
  toTile = s: if elem s specialTileTypes then toSpecialTile s else toAppTile s;

  # The "tile" type
  # Or maybe a freeform submodule is better?
  tileType = (types.attrsOf anything).addCheck (attrs: attrs ? tile-data);

  # The "coerced" tile type
  coercedTileType = types.coercedTo (types.either types.path types.str) toTile tileType;
in
types.nullOr (types.listOf coercedTileType);
```

This is a change from the existing code though, so maybe out of scope for this PR?

Reply:
> a) allow users to define custom tile attrsets
I think the type coercion looks nice, but is `addCheck` only available at the top level of `lib.types` ?

Reply:
> but is addCheck only available at the top level of lib.types ?

Yeah I misremembered the exact usage.

It should be `addCheck (attrsOf anything) (hasAttr "attrname")`

Ideally the type would have an overridden description too, or be a submodule, but I don't know how far you want to go with verbose nix code 😅

Reply:
@Enzime Any opinion on implementation options?